### ISSUE TITLE
Bug Rendering scene.visuals.Line with color as NumpyArray with alpha=0.0

### ISSUE THREAD

**Issue by alexjc** [vispy/vispy#588]
If you pass in a Numpy Array as the color parameter of a Line(), and some of the items have alpha =0.0, the whole line not only fades out but interpolates towards the origin.

This is in GL mode:
![screen_vispybuggl](https://cloud.githubusercontent.com/assets/445208/4607259/f4fa1c06-5247-11e4-9f49-a538f2bd6995.png)

This is in AGG mode:
![screen_vispybugagg](https://cloud.githubusercontent.com/assets/445208/4607258/f4dea566-5247-11e4-997d-303b57ae0ced.png)

**Comment by djhoese** [comment]
Coming back to this (after 5 years), this is what I get on a macbook with retina display if I change @alexjc's alpha setting to use `0.2` instead of `0`:

![image](https://user-images.githubusercontent.com/1828519/60917852-775dc800-a257-11e9-8498-77614ba6159e.png)


Setting it to 0 gives me the same weird results he was getting.

I noticed in the agg.frag shader that there is a line that explicitly throws away anything that is completely transparent:

```
    // If color is fully transparent we just discard the fragment
    if( v_color.a <= 0.0 ) {
        discard;
    }
```

**Comment by djhoese** [comment]
The part of the vertex shader that is relevant to alpha handling:

```
    // Thickness below 1 pixel are represented using a 1 pixel thickness
    // and a modified alpha
    v_color.a = min(v_linewidth, v_color.a);
    v_linewidth = max(v_linewidth, 1.0);

    // If color is fully transparent we just will discard the fragment anyway
    if( v_color.a <= 0.0 )
    {
        gl_Position = vec4(0.0,0.0,0.0,1.0);
        return;
    }
```

If I comment out the `if` statement then the effect goes away and the output is *I think* correct:

![image](https://user-images.githubusercontent.com/1828519/60918300-9d379c80-a258-11e9-9056-070fb137d7f1.png)

Can someone clarify whether or not the "overlap" brightness of the AGG lines is expected/desired?

**Comment by rougier** [comment]
Difficult to say with this example. Maybe a simple example with two broken lines would help. I know there are some corner cases with the agg shader but the example above seems to be different somehow.

**Comment by djhoese** [comment]
@rougier Yeah good point. I made a simpler example by reducing the number of lines shown on the canvas and changing the line coordinates to a sine wave with fewer points I can see the "brightening" effect:

![image](https://user-images.githubusercontent.com/1828519/60973092-29939f00-a2ed-11e9-9eac-118af09e0c2b.png)

If I change the hardcoded AGG line join type from "round" to "miter" I can see that the bright parts change shape/type so these are definitely line joins. Seeing as the vertex shader effects so much of the fragment shader, its not the easiest to debug.

@rougier Do you know where the original AGG code that's in vispy came from? Somewhere maybe I could look for updates to see if this kind of issue was fixed.

Is commenting out the `if` statement a good enough fix for at least this issue? I could include it in 0.6 and we can worry about this brightened line joins issue later. Or I could not make the fix and we address this later. Thoughts @larsoner @rougier @kmuehlbauer?

<details>

```
# -*- coding: utf-8 -*-
# vispy: gallery 30
# -----------------------------------------------------------------------------
# Copyright (c) Vispy Development Team. All Rights Reserved.
# Distributed under the (new) BSD License. See LICENSE.txt for more info.
# -----------------------------------------------------------------------------
"""
Demonstration of various features of Line visual.
"""

from __future__ import division

import sys
import numpy as np

from vispy import app, gloo, visuals
from vispy.visuals.transforms import STTransform, NullTransform

# vertex positions of data to draw
N = 45
pos = np.zeros((N, 2), dtype=np.float32)
pos[:, 0] = np.linspace(10, 390, N)
pos[:, 1] = np.sin(np.linspace(0, np.pi * 20, N)) * 40
# pos[:, 1] = np.random.normal(size=N, scale=20, loc=0)

# color array
color = np.ones((N, 4), dtype=np.float32)
color[:, 0] = np.linspace(0, 1, N)
color[:, 1] = color[::-1, 0]
# for i, c in enumerate(color):
#      if i % 5 == 0: c[3] = 0.0

# connection array
connect = np.empty((N-1, 2), np.int32)
connect[:, 0] = np.arange(N-1)
connect[:, 1] = connect[:, 0] + 1
connect[N//2, 1] = N//2  # put a break in the middle


class Canvas(app.Canvas):
    def __init__(self):
        app.Canvas.__init__(self, keys='interactive',
                            size=(800, 200))
        # Create several visuals demonstrating different features of Line
        self.lines = [
            # agg-method lines:
            # per-vertex color
            visuals.LineVisual(pos=pos, color=color, width=7., method='agg'),
            # visuals.LineVisual(pos=pos, color=color, method='agg'),

            # GL-method lines:
            visuals.LineVisual(pos=pos, color=color, method='gl'),
        ]
        counts = [0, 0]
        for i, line in enumerate(self.lines):
            # arrange lines in a grid
            tidx = (line.method == 'agg')
            x = 400 * tidx
            y = 140 * (counts[tidx] + 1)
            counts[tidx] += 1
            line.transform = STTransform(translate=[x, y])
            # redraw the canvas if any visuals request an update
            line.events.update.connect(lambda evt: self.update())

        self.texts = [visuals.TextVisual('GL', bold=True, font_size=24,
                                         color='w', pos=(200, 40)),
                      visuals.TextVisual('Agg', bold=True, font_size=24,
                                         color='w', pos=(600, 40))]
        for text in self.texts:
            text.transform = NullTransform()
        self.visuals = self.lines + self.texts
        self.show()

    def on_draw(self, event):
        gloo.clear('black')
        for visual in self.visuals:
            visual.draw()

    def on_resize(self, event):
        # Set canvas viewport and reconfigure visual transforms to match.
        vp = (0, 0, self.physical_size[0], self.physical_size[1])
        self.context.set_viewport(*vp)
        for visual in self.visuals:
            visual.transforms.configure(canvas=self, viewport=vp)


if __name__ == '__main__':
    win = Canvas()

    def update(ev):
        pos[:, 1] = np.random.normal(size=N, scale=30, loc=0)
        win.lines[0].set_data(pos)
        win.lines[3].set_data(pos)

    # timer = app.Timer()
    # timer.connect(update)
    # timer.start(1.0)

    if sys.flags.interactive != 1:
        app.run()

```

</details>

**Comment by larsoner** [comment]
+1 for incremental progress even if imperfect

**Comment by rougier** [comment]
I made the code and the shader for the AGG but I don't remember this kind of problem with line join. There are other problems for sure but not this specific one. Blend mode might be the culprit.

**Comment by almarklein** [comment]
The original issue (in the title) has been fixed in #1661. What remains is a brightening artifact at some of the joins in the line. I tried to reproduce this with the `basics/visuals/line.py` example, but I can't reproduce the artifact on my machine.

**Comment by djhoese** [comment]
Yeah I'm OK closing this. I'm having trouble reproducing the bright spots with current main and my example code from above (in the `Details` part of my comment). Let's close this and if someone has a more concrete example that is clearly wrong then please create a new issue and reference this one.
